File: modules/piu/PC/mac/piuFont.m
/*
 * Copyright (c) 2016-2017  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 * 
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 * 
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "piuPC.h"

static void PiuFontDelete(void* it);
static void PiuFontMark(xsMachine* the, void* it, xsMarkRoot markRoot);
static void PiuFontListMark(xsMachine* the, void* it, xsMarkRoot markRoot);

static xsHostHooks PiuFontHooks = {
	PiuFontDelete,
	PiuFontMark,
	NULL
};

static const xsHostHooks PiuFontListHooks = {
	NULL,
	PiuFontListMark,
	NULL
};

void PiuFontDelete(void* it)
{
}

PiuDimension PiuFontGetAscent(PiuFont* self)
{
	return c_ceil((*self)->ascent);
}

PiuDimension PiuFontGetHeight(PiuFont* self)
{
	return c_ceil((*self)->height);
}

PiuDimension PiuFontGetWidth(PiuFont* self, xsSlot* slot, xsIntegerValue offset, xsIntegerValue length)
{
	return c_ceil(PiuFontGetWidthSubPixel(self, slot, offset, length));
}

double PiuFontGetWidthSubPixel(PiuFont* self, xsSlot* slot, xsIntegerValue offset, xsIntegerValue length)
{
	xsMachine* the = (*self)->the;
	xsStringValue value = PiuToString(slot);
	value += offset;
	if (length < 0)
		length = c_strlen(value);
	CFStringRef string = CFStringCreateWithBytesNoCopy(NULL, (uint8_t*)value, length, kCFStringEncodingUTF8, false, kCFAllocatorNull);
    CFRange range = CFRangeMake(0, CFStringGetLength(string));
	CFMutableAttributedStringRef attributedString = CFAttributedStringCreateMutable(kCFAllocatorDefault, 0);
	CFAttributedStringReplaceString(attributedString, CFRangeMake(0, 0), string);
	CFAttributedStringSetAttribute(attributedString, range, kCTFontAttributeName, (*self)->fref);
    CTLineRef line = CTLineCreateWithAttributedString(attributedString);
    CGFloat width = CTLineGetOffsetForStringIndex(line, range.length, NULL);
  	CFRelease(line);
 	CFRelease(attributedString);
 	CFRelease(string);
 	return width;
}

void PiuFontMark(xsMachine* the, void* it, xsMarkRoot markRoot)
{
}

void PiuFontNew(xsMachine* the)
{
	PiuFont* self;
	xsResult = xsNewHostObject(NULL);
	xsSetHostChunk(xsResult, NULL, sizeof(PiuFontRecord));
	self = PIU(Font, xsResult);
	(*self)->reference = xsToReference(xsResult);
	(*self)->the = the;
	xsSetHostHooks(xsResult, &PiuFontHooks);
}

void PiuStyleLookupFont(PiuStyle* self)
{
	xsMachine* the = (*self)->the;
	xsResult = xsGet(xsGlobal, xsID_fonts);
	PiuFontList* fontList;
	if (xsTest(xsResult))
		fontList = PIU(FontList, xsResult);
	else {
		PiuFontListNew(the);
		xsSet(xsGlobal, xsID_fonts, xsResult);
		fontList = PIU(FontList, xsResult);
	}
	PiuFont* font = (*fontList)->first;
	while (font) {
		if (((*font)->family == (*self)->family)
				&& ((*font)->size == (*self)->size)
				&& ((*font)->weight == (*self)->weight)
				&& ((*font)->flags == ((*self)->flags & (piuStyleBits | piuStyleDecorationBits)))) {
			(*self)->font = font;
			return;
		}
		font = (*font)->next;
	}
	char buffer[256];
	xsStringValue name = xsName((*self)->family);
	if (!name) {
		NSFont *systemFont = [NSFont systemFontOfSize:(*self)->size];
		NSString* string = systemFont.fontName;
		name = (xsStringValue)[string UTF8String];
	}
	if (name)
		c_strcpy(buffer, name);
	else
		c_strcpy(buffer, "undefined");
	if (((*self)->flags & (piuStyleCondensedBit | piuStyleItalicBit)) || ((*self)->weight)) {
		c_strcat(buffer, "-"); 
		if ((*self)->flags & piuStyleCondensedBit)
			c_strcat(buffer, "Condensed"); 
		switch ((*self)->weight) {
			case 1: c_strcat(buffer, "Ultralight"); break;
			case 2: c_strcat(buffer, "Thin"); break;
			case 3: c_strcat(buffer, "Light"); break;
			case 5: c_strcat(buffer, "Medium");break;
			case 6: c_strcat(buffer, "Semibold"); break;
			case 7: c_strcat(buffer, "Bold");break;
			case 8: c_strcat(buffer, "Heavy"); break;
			case 9: c_strcat(buffer, "Black");break;
		}
		if ((*self)->flags & piuStyleItalicBit)
			c_strcat(buffer, "Italic"); 
	}
	else
		c_strcat(buffer, "-Regular"); 
	//fprintf(stderr, "%x %s %d %d -> %s\n", (*self)->flags, name, (*self)->size, (*self)->weight, buffer);
	
	CFStringRef string = CFStringCreateWithBytesNoCopy(NULL, (uint8_t*)buffer, c_strlen(buffer), kCFStringEncodingUTF8, false, kCFAllocatorNull);
	CTFontDescriptorRef descriptor = CTFontDescriptorCreateWithNameAndSize(string, (*self)->size);
  
  CTFontRef fref = CTFontCreateWithFontDescriptor(descriptor, 0.0, NULL);
  if ((*self)->flags & piuStyleItalicBit) {
  	CTFontRef frefItalic = CTFontCreateCopyWithSymbolicTraits(fref, 0.0, NULL, kCTFontItalicTrait, kCTFontItalicTrait);
  	CFRelease(fref);
  	fref = frefItalic;
  }
  CFRelease(descriptor);
	CFRelease(string);
	
	CGFloat ascent = CTFontGetAscent(fref);
	CGFloat descent = CTFontGetDescent(fref);
	CGFloat leading = CTFontGetLeading(fref);
	
	PiuFontNew(the);
	font = PIU(Font, xsResult);
    (*font)->next = (*fontList)->first;
    (*fontList)->first = font;
	
  (*font)->flags = (*self)->flags & (piuStyleBits | piuStyleDecorationBits);
	(*font)->family = (*self)->family;
	(*font)->size = (*self)->size;
	(*font)->weight = (*self)->weight;
	
	(*font)->fref = fref;
	(*font)->ascent = ascent;
	(*font)->height = ascent + descent + leading;
	
	(*self)->font = font;
}

void PiuFontListMark(xsMachine* the, void* it, xsMarkRoot markRoot)
{
	PiuFontList self = it;
	PiuFont* font = self->first;
	while (font) {
		PiuMarkHandle(the, font);
		font = (*font)->next;
	}
}

void PiuFontListNew(xsMachine* the)
{
	PiuFontList* self;
	xsResult = xsNewHostObject(NULL);
	xsSetHostChunk(xsResult, NULL, sizeof(PiuFontListRecord));
	self = PIU(FontList, xsResult);
	(*self)->reference = xsToReference(xsResult);
	xsSetHostHooks(xsResult, &PiuFontListHooks);
}

File: modules/piu/Pebble/piuFont.c
#include "piuPebble.h"

static void PiuFontDelete(void* it);
static void PiuFontMark(xsMachine* the, void* it, xsMarkRoot markRoot);
static void PiuFontListDelete(void* it);
static void PiuFontListMark(xsMachine* the, void* it, xsMarkRoot markRoot);

static xsHostHooks PiuFontHooks = {
	PiuFontDelete,
	PiuFontMark,
	NULL
};

static const xsHostHooks PiuFontListHooks = {
	PiuFontListDelete,
	PiuFontListMark,
	NULL
};

void PiuFontDelete(void* it)
{
}

PiuDimension PiuFontGetAscent(PiuFont* self)
{
	return (*self)->ascent;
}

PiuDimension PiuFontGetHeight(PiuFont* self)
{
	return (*self)->height;
}

PiuDimension PiuFontGetWidth(PiuFont* self, xsSlot* string, xsIntegerValue offset, xsIntegerValue length)
{
	xsMachine* the = (*self)->the;
	xsStringValue text = PiuToString(string);
	PiuDimension width = 0;
	text += offset;
	if ((*self)->gfont) {
		char tmp[100];

		GContext *ctx = app_state_get_graphics_context();
		if ((-1 != length) && text[length]) {
			if (length >= (xsIntegerValue)sizeof(tmp))
				length = sizeof(tmp) - 1;

			c_memmove(tmp, text, length);
			tmp[length] = 0;
			text = tmp;
		}

		GSize size = graphics_text_layout_get_max_used_size(
					ctx, text,
				(*self)->gfont, GRect(0, 0, 10000, 100),
					GTextOverflowModeTrailingEllipsis,
					GTextAlignmentLeft, NULL);


		width = size.w;
	}
	else {
		xsIntegerValue character = 0;
		CFEGlyph glyph;
		CFESetFontData(gCFE, (*self)->buffer, (*self)->bufferSize);
		CFESetFontSize(gCFE, (*self)->size);
		while (length) {
			xsStringValue formerText = text;
	#if MODDEF_CFE_KERN
			xsIntegerValue formerCharacter = character;
	#endif
			text = fxUTF8Decode(text, &character);
			if (character <= 0)
				break;
			length -= text - formerText;
			glyph = CFEGetGlyphFromUnicode(gCFE, character, 0);
			if (!glyph) {
				character = '?';
				glyph = CFEGetGlyphFromUnicode(gCFE, character, 0);
				if (!glyph)
					continue;
			}
			width += glyph->advance;
	#if MODDEF_CFE_KERN
			if (formerCharacter)
				width += CFEGetKerningOffset(gCFE, formerCharacter, character);
	#endif
		}
	}
	return width;
}


void PiuFontMark(xsMachine* the, void* it, xsMarkRoot markRoot)
{
	PiuFont self = it;
	PiuMarkHandle(the, self->texture);
}

void PiuFontNew(xsMachine* the)
{
	PiuFont* self;
	xsResult = xsNewHostObject(NULL);
	xsSetHostChunk(xsResult, NULL, sizeof(PiuFontRecord));
	self = PIU(Font, xsResult);
	(*self)->reference = xsToReference(xsResult);
	(*self)->the = the;
	xsSetHostHooks(xsResult, &PiuFontHooks);
}

void PiuStyleLookupFont(PiuStyle* self)
{
	xsMachine* the = (*self)->the;
	void* archive = NULL;
	PiuFontList* fontList;
	PiuFont* font = NULL;
	xsStringValue name;
	char path[256];
	void* buffer;
	size_t bufferSize;
	int32_t ascent, descent, leading;
	
	if ((*self)->archive) {
		xsResult = xsReference((*self)->archive);
		archive = xsGetHostData(xsResult);
	}
	xsResult = xsGet(xsGlobal, xsID_fonts);
	if (xsTest(xsResult))
		fontList = PIU(FontList, xsResult);
	else {
		gCFE = CFENew();
		PiuFontListNew(the);
		xsSet(xsGlobal, xsID_fonts, xsResult);
		fontList = PIU(FontList, xsResult);
	}
	font = (*fontList)->first;
	while (font) {
		if (((*font)->family == (*self)->family)
				&& ((*font)->size == (*self)->size)
				&& ((*font)->weight == (*self)->weight)
				&& ((*font)->flags == ((*self)->flags & piuStyleBits))) {
			(*self)->font = font;
			return;
		}
		font = (*font)->next;
	}
	
	name = xsName((*self)->family);
	if (name)
		c_strcpy(path, name);
	else
		c_strcpy(path, "undefined");
	if (((*self)->flags & (piuStyleCondensedBit | piuStyleItalicBit)) || ((*self)->weight) || ((*self)->size)) {
		xsIntegerValue flag = 0;
		c_strcat(path, "-");
		if ((*self)->flags & piuStyleCondensedBit)
			c_strcat(path, "Condensed");
		else
			flag |= 1;
		switch ((*self)->weight) {
			case 1: c_strcat(path, "Ultralight"); break;
			case 2: c_strcat(path, "Thin"); break;
			case 3: c_strcat(path, "Light"); break;
			case 5: c_strcat(path, "Medium");break;
			case 6: c_strcat(path, "Semibold"); break;
			case 7: c_strcat(path, "Bold");break;
			case 8: c_strcat(path, "Heavy"); break;
			case 9: c_strcat(path, "Black");break;
			default: flag |= 2; break;
		}
		if ((*self)->flags & piuStyleItalicBit)
			c_strcat(path, "Italic"); 
		else
			flag |= 4;
		if (flag == 7)
			c_strcat(path, "Regular");
	}
	
	PiuFontNew(the);
	font = PIU(Font, xsResult);
	(*font)->gfont = modFindPebbleFont(path, (*self)->size, &ascent, &descent, &leading);
	if ((*font)->gfont == NULL) {
		if ((*self)->size) {
			xsIntegerValue length = c_strlen(path) + 1;
			c_strcat(path, "-");
			fxIntegerToString(NULL, (*self)->size, path + length, sizeof(path) - length);
		}
		name = path + c_strlen(path);
		c_strcpy(name, ".fnt");
		buffer = (uint8_t *)fxGetResource(the, archive, path, &bufferSize);
		if (!buffer)
			xsURIError("font not found: %s", path);
		(*font)->buffer = buffer;
		(*font)->bufferSize = bufferSize;
		CFESetFontData(gCFE, buffer, bufferSize);
		CFESetFontSize(gCFE, (*self)->size);
		CFEGetFontMetrics(gCFE, &ascent, &descent, &leading);
		
		(*font)->next = (*fontList)->first;
		(*fontList)->first = font;
			
		c_strcpy(name, ".png");
		xsResult = xsNewObject();
		if (archive)
			xsDefine(xsResult, xsID_archive, xsReference((*self)->archive), xsDefault);
		xsDefine(xsResult, xsID_path, xsString(path), xsDefault);
		xsResult = xsNew1(xsGlobal, xsID_Texture, xsResult);
		(*font)->texture = PIU(Texture, xsResult);
	}
	else {
		(*font)->next = (*fontList)->first;
		(*fontList)->first = font;
 	}
   
	(*font)->flags = (*self)->flags & piuStyleBits;
	(*font)->family = (*self)->family;
	(*font)->size = (*self)->size;
	(*font)->weight = (*self)->weight;
	
	(*font)->ascent = ascent;
	(*font)->height = ascent + descent + leading;
	
	(*self)->font = font;
}

void PiuFontListDelete(void* it)
{
	if (gCFE) {
		CFEDispose(gCFE);
		gCFE = C_NULL;
	}
}

void PiuFontListMark(xsMachine* the, void* it, xsMarkRoot markRoot)
{
	PiuFontList self = it;
	PiuFont* font = self->first;
	while (font) {
		PiuMarkHandle(the, font);
		font = (*font)->next;
	}
}

void PiuFontListNew(xsMachine* the)
{
	PiuFontList* self;
	xsResult = xsNewHostObject(NULL);
	xsSetHostChunk(xsResult, NULL, sizeof(PiuFontListRecord));
	self = PIU(FontList, xsResult);
	(*self)->reference = xsToReference(xsResult);
	xsSetHostHooks(xsResult, &PiuFontListHooks);
}
